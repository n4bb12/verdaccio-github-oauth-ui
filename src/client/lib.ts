/**
 * Returns `?a=b&c` as `{ a: b, c: true }`.
 */
export function parseQueryParams() {
  return (location.search || "?")
    .substring(1)
    .split("&")
    .filter(kv => kv)
    .map(kv => kv.split("=").concat(["true"]))
    .reduce((obj: any, pair) => {
      obj[pair[0]] = decodeURIComponent(pair[1])
      return obj
    }, {})
}

export function clearQueryParams() {
  history.replaceState(null, "", location.pathname)
}
