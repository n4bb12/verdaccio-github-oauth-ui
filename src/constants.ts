export const pluginName = "github-oauth-ui"
export const publicRoot = __dirname + "/public"
export const staticPath = "/-/static/" + pluginName
export const authorizePath = "/-/oauth/authorize"
export const callbackPath = "/-/oauth/callback"
