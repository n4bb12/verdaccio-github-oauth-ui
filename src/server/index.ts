import { Plugin } from "./plugin/Plugin"
import dotenv from "dotenv"

dotenv.config()

// plugins must be a default export
export default Plugin
