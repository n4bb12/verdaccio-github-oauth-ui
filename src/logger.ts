import chalk from "chalk"

import { pluginName } from "./constants"

const prefix = chalk.blue(`[${pluginName}]`)

// tslint:disable: no-console
export const logger = {
  log: console.log.bind(console, prefix),
  error: console.error.bind(console, prefix),
}
