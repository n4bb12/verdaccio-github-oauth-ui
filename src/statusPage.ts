import { pluginKey } from "./constants"

const styles = `
html,
body {
  padding: 0;
  margin: 0;
  height: 100%;
  background-color: #e0e0e0;
  color: #24292F;
  font-family: Helvetica, sans-serif;
  position: relative;
  text-align: center;
}
.wrap {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
}
a {
  color: #3f51b5;
}
.img {
  filter: drop-shadow(0 0.5rem 0.5rem #24292F80);
  width: 100px;
}
`
  .trim()
  .replace(/\s+/g, " ")

const logo = `<svg class="img" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100"><defs/><defs><filter id="a" width="139.6%" height="140.4%" x="-20%" y="-12%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feComposite in="shadowBlurOuter1" in2="SourceAlpha" operator="out" result="shadowBlurOuter1"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter><filter id="c" width="167.9%" height="272.7%" x="-34%" y="-50%" filterUnits="objectBoundingBox"><feOffset dy="4" in="SourceAlpha" result="shadowOffsetOuter1"/><feGaussianBlur in="shadowOffsetOuter1" result="shadowBlurOuter1" stdDeviation="2.5"/><feColorMatrix in="shadowBlurOuter1" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.0906646286 0"/></filter><path id="b" d="M48 17L33 47h-9L0 0h15l13 25 5-8h15z"/><path id="d" d="M35 11h-7V9h8l2-3h-6V4h7l1-2h-4V0h20l-6 11H35z"/></defs><g fill="none" fill-rule="evenodd"><rect width="100" height="100" fill="#000" rx="37"/><g transform="translate(22 30)"><use fill="#000" filter="url(#a)" xlink:href="#b"/><use fill="#FFF" fill-opacity=".6" xlink:href="#b"/></g><g transform="translate(22 30)"><use fill="#000" filter="url(#c)" xlink:href="#d"/><use fill="#FFF" xlink:href="#d"/></g><path fill="#FFF" d="M55 77h-9L22 30h15l21 41z"/></g></svg>`

export function buildStatusPage(body: string, withBackLink: boolean) {
  const backButton = withBackLink ? `<p><a href="/">Go back</a></p>` : ""

  return `<!DOCTYPE html>
<html lang="en">
  <head>
    <title>${pluginKey}</title>
    <style>${styles}</style>
  </head>
  <body>
    <div class="wrap">
      ${logo}
      ${body}
      ${backButton}
    </div>
  </body>
</html>`
}

export function buildErrorPage(error: any, withBackLink: boolean) {
  return buildStatusPage(
    `<h1>Sorry :(</h1>
    <p>${error?.message || error}</p>`,
    withBackLink,
  )
}
