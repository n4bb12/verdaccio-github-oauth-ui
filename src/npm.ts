import { execSync } from "child_process"
import minimist from "minimist"
import { URL } from "url"

import { logger } from "./logger"

let npmConfig: any

function parseCliArgs() {
  return minimist(process.argv.slice(2))
}

function runCommand(command: string) {
  logger.log(`Running command: ${command}`)
  return execSync(command)
}

function getNpmConfig() {
  if (!npmConfig) {
    npmConfig = JSON.parse(runCommand("npm config list --json").toString())
  }
  return npmConfig
}

function removeTrailingSlash(input: string) {
  return input.trim().replace(/\/?$/, "")
}

export function getRegistryUrl() {
  const cliArgs = parseCliArgs()

  const registry = cliArgs.registry || getNpmConfig().registry

  return removeTrailingSlash(registry)
}

export function getNpmConfigFile() {
  return getNpmConfig().userconfig
}

export function getNpmSaveCommands(registry: string, token: string) {
  const url = new URL(registry)
  const baseUrl = url.host + url.pathname

  return [
    `npm config set //${baseUrl}:always-auth true`,
    `npm config set //${baseUrl}:_authToken "${token}"`,
  ]
}

export function saveNpmToken(token: string) {
  const registry = getRegistryUrl()
  const commands = getNpmSaveCommands(registry, token)

  commands.forEach((command) => runCommand(command))
}
