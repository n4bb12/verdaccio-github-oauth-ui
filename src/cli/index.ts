import express from "express"
import open from "open"

import { buildStatusPage } from "../statusPage"
import { getConfigFile, getRegistry, save } from "./npm"
import { printUsage } from "./usage"

const registry = getRegistry()

if (registry.includes("registry.npmjs.org")) { // lgtm [js/incomplete-url-substring-sanitization]
  printUsage()
  process.exit(1)
}

open(registry + "/oauth/authorize")

const successPage = buildStatusPage(`
  <h1>All done!</h1>
  <p>We've updated your npm configuration.</p>
  <p><code>${getConfigFile()}</code></p>
`)

const server = express()
  .get("/", (req, res) => {
    const token = decodeURIComponent(req.query.token)
    save(registry, token)
    res.setHeader("Content-Type", "text/html")
    res.send(successPage)
    server.close()
  })
  .listen(8239)
