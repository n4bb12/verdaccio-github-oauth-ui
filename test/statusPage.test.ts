import { buildStatusPage } from "src/statusPage"

describe("Shared", () => {
  describe("buildStatusPage", () => {

    it("matches the snapshot", () => {
      expect(buildStatusPage("custom test body")).toMatchSnapshot()
    })

  })
})
