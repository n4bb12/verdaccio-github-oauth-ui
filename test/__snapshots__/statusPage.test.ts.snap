// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shared buildStatusPage matches the snapshot 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <title>github-oauth-ui</title>
    <style>
      html,
      body {
        padding: 0;
        margin: 0;
        height: 100%;
        background-color: #e0e0e0;
        color: #24292F;
        font-family: Helvetica, sans-serif;
        position: relative;
        text-align: center;
      }
      .wrap {
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
      }
      a {
        color: #3f51b5;
        text-transform: uppercase;
      }
      img {
        filter: drop-shadow(0 0.5rem 0.5rem #24292F80);
      }
    </style>
  </head>
  <body>
    <div class=\\"wrap\\">
      <img src=\\"https://verdaccio.org/docs/assets/logo/symbol/svg/verdaccio-blackwhite.svg\\" alt=\\"verdaccio-blackwhite.svg\\" />
      custom test body
    </div>
  </body>
</html>"
`;
